### PR TITLE
Feature/tests microservices

### PR DESCRIPTION
**Pull request by RoykeeOM2**
This PR includes tests for the implementation files.

1. ImportHtmlService.spec.ts
2. Microservice.spec.ts
3. ProxyFromMicroservice.spec.ts
4. ServiceCallFromMicroservice.spec.ts

### CODE REVIEW

File: packages/scalecube-microservice/tests/implementations/ProxyFromMicroservice.spec.ts [hunk 1]
```diff
@@ -100,7 +112,18 @@ describe('Test creating proxy from microservice', () => {
     }, 1000);
   });
 
-  it('Throw error message when creating proxy with missing serviceDefinition', () => {
+  test(`
+    Scenario: Fail to create proxy, missing serviceDefinition
+      Given   a service with definition and reference
+      And     definition and reference comply with each other
```
Comment (IdoJS) [inline packages/scalecube-microservice/tests/implementations/ProxyFromMicroservice.spec.ts]:
```suggestion
      And     definition and reference **are not** comply with each other
```

Comment (RoykeeOM2) [reply packages/scalecube-microservice/tests/implementations/ProxyFromMicroservice.spec.ts]: I have fixed, using different word choice ;)

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 2]
```diff
@@ -31,7 +31,18 @@ describe('Test the creation of Microservice', () => {
...
 
   describe('Test creating microservice from function constructor', () => {
-    test('MethodRegistry throws an error when method reference is not a function', () => {
+    test(`
+      Scenario: Fail to register a service, reference does not match definition
+        Given   a service with definition and reference
+        |service          |definition            |reference  |
+        |greetingService  |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
+        |                 |empty : null          |                      |
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]:
```suggestion
        |                 |hello2 : RequestResponse  |                      |
```

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 3]
```diff
@@ -62,7 +73,19 @@ describe('Test the creation of Microservice', () => {
...
+      Scenario: Fail to register a service, reference method is not a function
+        Given   a service with definition and reference
+        And     definition and reference comply with each other
+        |service          |definition            |reference             |
+        |greetingService  |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
+        |                 |empty: null           |empty: null           |
+        # reference has a method that is not a function
+        When    creating a Microservice with the service
+        Then    exception will occur
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: Add a line:
`Then  a Microservice will not be created`
`And    exception...`

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 4]
```diff
@@ -62,7 +73,19 @@ describe('Test the creation of Microservice', () => {
...
+        And     Invalid method reference for GreetingService/empty
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: Combine lines 86&87

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 5]
```diff
@@ -31,7 +31,18 @@ describe('Test the creation of Microservice', () => {
...
+      Scenario: Fail to register a service, reference does not match definition
+        Given   a service with definition and reference
+        |service          |definition            |reference  |
+        |greetingService  |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
+        |                 |empty : null          |                      |
+        # definition has a method that is not contained in the reference
+        When    creating a Microservice with the service
+        Then    exception will occur
+        And     Invalid method reference for GreetingService/empty
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]:
```suggestion
        And     service method  ‘empty’ missing in the serviceDefinition`
```

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 6]
```diff
@@ -100,14 +145,40 @@ describe('Test the creation of Microservice', () => {
...
-      test('Test REQUEST_RESPONSE', () => {
+      test(`
+      Scenario: Invoke registered serviceS from proxy (RequestResponse)
+        Given   a service with definition and reference
+        And     definition and reference comply with each other
+        |service          |definition            |reference             |
+        |greetingService1 |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
+        |greetingService2 |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: Change to:
|service          |definition            |reference             |
|greetingService1 |greet$: RequestStream |greet$: RequestStream |
|greetingService2 |hello: RequestResponse|hello: RequestResponse|

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 7]
```diff
@@ -100,14 +145,40 @@ describe('Test the creation of Microservice', () => {
...
-      test('Test observable', (done) => {
+      test(`
+      Scenario: Subscribe to a service from proxy (RequestStream)
+        Given   a service with definition and reference
+        And     definition and reference comply with each other
+        |service          |definition            |reference             |
+        |greetingService1 |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
+        |greetingService2 |hello: RequestResponse|hello: RequestResponse|
+        |                 |greet$: RequestStream |greet$: RequestStream |
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: Change to:
|service |definition |reference |
|greetingService1 |greet$: RequestStream |greet$: RequestStream |
|greetingService2 |hello: RequestResponse|hello: RequestResponse|

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 8]
```diff
@@ -120,7 +191,16 @@ describe('Test the creation of Microservice', () => {
...
-    test('Access endpoint from other discovery with remoteCall', () => {
+    test(`
+      Scenario: Invoke remote service 
+        Given   two valid Microservices under the same seedAddress
+        | Microservice | reference| definition  |
+        | ms1          | s1       |d1           |
+        | ms2          | s2       |d2           | 
+        When    creating a proxy to ms1 with definition of d2
+        And     proxy invokes s2
+        Then    valid response is received
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]:
```suggestion
        Then    remoteCall is performed and valid response is received
```

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts [hunk 9]
```diff
@@ -139,5 +219,19 @@ describe('Test the creation of Microservice', () => {
...
     });
+
+    test(`
+      Scenario: Fail to invoke a remote service 
+        Given   two valid Microservices under the same seedAddress
+        | Microservice | reference| definition  |
+        | ms1          | s1       |d1           |
+        | ms2          | s2       |d2           | 
+        When    creating a proxy to ms1 with definition of d3
+        And     proxy tries to invoke s2
```
Comment (RoykeeOM2) [inline packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]:
```suggestion
        And     remoteCall is performed and proxy tries to invoke s2
```

Comment (RoykeeOM2) [reply packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: Add a scenario where we have definition 'Foo' and reference 'Foo' & 'Ro' to a single ms.
The result will be that ms will be created without 'Ro'.
`
      Scenario: Fail to register a service, 
        Given   a service with definition and reference
        |service          |definition            |reference  |
        |greetingService  |             |hello: RequestResponse|
        |                 |greet$: RequestStream |greet$: RequestStream |

        # reference has a method that is not contained in the definition
        When    creating a Microservice with the service
        Then    Microservice will successfully created
        But Microservice instance won't contain hello method
      `

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts (hunk 3 above)
Comment (RoykeeOM2) [reply packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: @IdoJS can you check for the relevant error that will be emitted in that case? I would like to write in the test the response as I did in other tests.

File: packages/scalecube-microservice/tests/implementations/Microservice.spec.ts (hunk 4 above)
Comment (RoykeeOM2) [reply packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: @IdoJS can you check for the relevant error that will be emitted in that case? I would like to write in the test the response as I did in other tests.

Comment (IdoJS) [reply packages/scalecube-microservice/tests/implementations/Microservice.spec.ts]: @RoykeeOM2 
please fix the table.
the empty in the definition should be requestResponse And not null

File: packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts [hunk 10]
```diff
@@ -73,15 +78,29 @@ const remoteResponse = ({
...
-          socketConnect.then(flowableNext, flowableError); // Single type
+          socketConnect.subscribe({
+            onComplete: flowableNext,
+            onError: flowableError,
+          }); // Single type
           break;
 
         case ASYNC_MODEL_TYPES.REQUEST_STREAM:
-          socketConnect.subscribe(flowableNext, flowableError, () => observer.complete()); // Flowable type
+          const max = socketConnect._max;
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts]: is it the value that is provided by Rsocket lib itself?

File: packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
+import { ASYNC_MODEL_TYPES, Microservices } from '../../../src';
+import { applyPostMessagePolyfill } from '../../mocks/utils/PostMessageWithTransferPolyfill';
+import { applyMessageChannelPolyfill } from '../../mocks/utils/MessageChannelPolyfill';
+import { catchError } from 'rxjs/operators';
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts]: It's recommended to put external modules imports before the internal imports

File: packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    When      remoteService emit error
+    Then      proxy receive the error message
+  `, (done) => {
+        expect.assertions(1);
+
+        const proxy = sender.createProxy({ serviceDefinition: receiverServiceDefinition });
+        proxy
+          .greet$(['Me'])
+          .pipe(
+            catchError((error: any) => {
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts]: better not to do it in **catchError**, but do it in error callback itself (the second argument of **subscribe** method)

File: packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+              expect(error).toMatchObject(new Error(errorMessage));
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts]: toEqual will also work fine

File: packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+              | method                | asyncModel definition | asyncModel reference|
+              | incorrectAsyncModel$  | requestStream         | requestResponse     |
+    When      remoteService return value
+    Then      proxy receive the value
+  `, (done) => {
+        expect.assertions(1);
+        const proxy = sender.createProxy({ serviceDefinition: receiverServiceDefinition });
+        proxy.incorrectAsyncModel$().subscribe((res: any) => {
+          expect(res).toMatchObject({ emptyMessage });
+          done();
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/Errors/remoteServiceCall.spec.ts]: you don't have to call done here - test should wait for the emit, if you have specified the amount of expectations

File: packages/scalecube-microservice/tests/integration/Errors/serverTransportProvider.spec.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    proxy
+      .greet$(['Me'])
+      .pipe(
+        catchError((error: any) => {
+          expect(error).toMatchObject(new Error(errorMessage));
+          done();
+          return [];
+        })
+      )
+      .subscribe();
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/Errors/serverTransportProvider.spec.ts]: also error callback

**Pull request merged**